### ISSUE TITLE
System Test Failures (core)

### ISSUE THREAD

**Issue by loganmc10** [simple64/simple64#284]
```
Core: IS64: Test 'cart: Read16' failed: a == b expected, but a=17767 b=35243 (hex: a=0x4567 b=0x89ab). Reading 16 bit from cart[1]
Core: IS64: Test 'cart: Read8' failed: a == b expected, but a=69 b=137 (hex: a=0x45 b=0x89). Reading 8 bit from cart[2]
Core: IS64: Test 'cart-writing: Write8 (with offset), Read32' failed: a == b expected, but a=12189696 b=1455030272 (hex: a=0xba0000 b=0x56ba0000). Reading first time from cart after writing
Core: IS64: Test 'pifram: SH (offset 0)' failed: a == b expected, but a=836912487 b=836894720 (hex: a=0x31e24567 b=0x31e20000). Reading 32 bit from PIFRAM[0]
Core: IS64: Test 'pifram: SH (offset 2)' failed: a == b expected, but a=836907426 b=1111568802 (hex: a=0x31e231a2 b=0x424131a2). Reading 32 bit from PIFRAM[0]
Core: IS64: Test 'pifram: SB' failed: a == b expected, but a=829960610 b=1450704896 (hex: a=0x317831a2 b=0x56780000). Reading 32 bit from PIFRAM[0]
```

**Comment by DarthSidious666** [comment]
That's interesting.
It's expecting a match but finding incorrect addresses.
I think that the cart errors match up with my assumption about the game timing not being read from the cart.

Also looks like some of the opcodes are implemented incorrectly and jumping to the wrong address or reading the wrong address.

Using build fa634d0.
I ran the n64-systemtest rom and I get the following:

In 4MB standard:
Nothing happens with the display.

```
Core: IS64: 
Heap range: 80199ea0Heap range: 80199ea0 to 80200000
Heap range: Heap range: 80199ea0 to 80200000
Heap range: 80199ea0 to 80200000
Heap range: 80199ea0 to 80200000
Heap range: 80199ea0 to 80200000
Core: IS64: Heap range: ff80ff80 to ff84ff84
Core: IS64: n64-systemtest Version Heap range: 80199ea0 to Heap range: 80199ea0 to 80200000
Core: IS64: n64-systemtest Version 1.0.14
Core Error: NI() @ 0x800F6DD4
Core Error: opcode not implemented: 800F6DD4:40217000
```

In 8MB expanded:
Display flashes black for a second.

```
Core: IS64: Test 'StartupTest' failed: a == b expected, but a=451683 b=1879499875 (hex: a=0x6e463 b=0x7006e463). Initial COP0 Config
Core: IS64: 
Core Error: NI() @ 0x800E6FD8
Core Error: opcode not implemented: 800E6FD8:40A22000
```

**Comment by loganmc10** [comment]
I've just updated the main branch with some fixes. I've updated the original issue with the remaining failures.

I haven't included failures related to 64-bit addressing or exception handling, maybe one day, but those shouldn't affect the compatibility of any games, so I'm going to forget about them for now.

The remaining issues are related to addressing alignment oddities in the PI and PIF.
